### PR TITLE
update System.CommandLine to the latest version

### PR DESCRIPTION
**Pull request by adamsitnik**
- stop using System.CommandLine.NamingConventionBinder as it's reflection based and won't be supported in the long term

- stop using ancient System.CommandLine.Experimental

Contributes to https://github.com/dotnet/sdk/pull/33157

### CODE REVIEW

File: NuGet.config [hunk 1]
```diff
@@ -11,6 +11,7 @@
     <add key="dotnet8" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json" />
     <add key="dotnet8-transport" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8-transport/nuget/v3/index.json" />
     <add key="dotnet-public" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json" />
+    <add key="dotnet-libraries" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-libraries/nuget/v3/index.json" />
```
Comment (jaredpar) [inline NuGet.config:14]: Nine different dotnet feeds now ... I feel like we're making NuGet unnecessarily complicated.

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs [hunk 2]
```diff
@@ -32,7 +30,7 @@
 WindowsErrorReporting.SetErrorModeOnWindows();
 
 var parser = CreateCommandLineParser();
-return await parser.InvokeAsync(args);
+return await parser.Parse(args).InvokeAsync(CancellationToken.None);
```
Comment (jasonmalinowski) [inline src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs:33]: What cancellation token should be going in here? Was the old behavior listening to Ctrl+C to raise the token or not?

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs [hunk 3]
```diff
@@ -121,60 +119,56 @@
...
         Description = "The name of the pipe used to connect to a remote process (if one exists).",
-        IsRequired = false,
+        Required = false,
     };
 
-    var logLevelOption = new Option<LogLevel>("--logLevel", description: "The minimum log verbosity.", parseArgument: result =>
+    var logLevelOption = new CliOption<LogLevel>("--logLevel")
     {
-        var value = result.Tokens.Single().Value;
-        return !Enum.TryParse<LogLevel>(value, out var logLevel)
```
Comment (jasonmalinowski) [inline src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs:140]: @dibarbet Were enums just not supported on the version we were on or was this necessary for some special reason?

File: src/Features/Lsif/Generator/Program.cs [hunk 4]
```diff
@@ -27,20 +25,41 @@ internal static class Program
...
+                    project: parseResult.GetValue(project),
+                    compilerInvocation: parseResult.GetValue(compilerInvocation),
+                    binLog: parseResult.GetValue(binLog),
+                    output: parseResult.GetValue(output),
+                    outputFormat: parseResult.GetValue(outputFormat),
+                    log: parseResult.GetValue(log),
+                    cancellationToken);
+            });
+
+            return generateCommand.Parse(args).InvokeAsync(CancellationToken.None);
```
Comment (jasonmalinowski) [inline src/Features/Lsif/Generator/Program.cs:62]: Same question as earlier: what should go here for the token?

**Review by jasonmalinowski — APPROVED** [review]
Generally looks OK, some questions about cancellation tokens but that could be done as a follow-up.

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs (hunk 2 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs:33]: The docs show `InvokeAsync` with no cancellationtoken.  Might be one of our analyzers telling it to pass a cancellation token?
https://learn.microsoft.com/en-us/dotnet/standard/commandline/handle-termination

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs (hunk 3 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs:140]: I honestly don't remember.  If this works with VSCode then I'm good with it.

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs (hunk 2 above)
Comment (adamsitnik) [reply src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs:33]: Whatever token we pass here, S.CL is going to create a source linked to it:

https://github.com/dotnet/command-line-api/blob/8f2ada7fbaab300dc3fcaa9141d1d69340e3b4cb/src/System.CommandLine/Invocation/InvocationPipeline.cs#L19

and create it's own token out of it, then request the cancellation when its Ctrl+C is pressed

https://github.com/dotnet/command-line-api/blob/8f2ada7fbaab300dc3fcaa9141d1d69340e3b4cb/src/System.CommandLine/Invocation/ProcessTerminationHandler.cs#L85

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs (hunk 3 above)
Comment (adamsitnik) [reply src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs:140]: enums are now supported OOTB, I can not speak about the past as I've been involved in this project only for a year and a half

File: src/Features/Lsif/Generator/Program.cs (hunk 4 above)
Comment (adamsitnik) [reply src/Features/Lsif/Generator/Program.cs:62]: > Same question as earlier: what should go here for the token?

If this is `Main`, then `None`.

If this was any other async method that would have it's own token, then that token.

**Pull request merged**